qemu/tests/driver_update.py

```diff
@@ -0,0 +1,148 @@
+            try:
+                vm.copy_files_from(log_file, result_dir)
+            except FileTransferNotFoundError:
+                logging.error("Didn't find file %s in guest" % log_file)
+
+    def get_driver_from_cdrom(driver_name, session):
+        """
+        Get driver path from virtio-win.iso, and driver version from inf file
+ 
+        :param driver_name: name of the driver
+        return: driver path and driver version
+        """
+        error.context("Get driver driver path and driver version"
+                       " from virtio-win iso", logging.info)
+        device_key = params.get("device_key", "VolumeName like '%virtio-win%'")
+        guest_name_alias = params.get("guest_alias")
+        guest_name = params.get("guest_name")
+        guest_list = {}
+        for i in guest_name_alias.split(","):
+            guest, guest_alias = i.split(":")
+            guest_list[guest] = guest_alias
+        guest_n = guest_list[guest_name]
+        virtio_win_disk = utils_misc.get_win_disk_vol(session, device_key) 
+        if not virtio_win_disk:
+            raise error.Error("Didn't fine virtio-win iso in guest")
+
+        driver_path = "%s:\%s\%s" % (virtio_win_disk, driver_name, guest_n)
+        cmd = "dir /B %s\*.inf" % driver_path
+        status, output = session.cmd_status_output(cmd) 
+        if status:
+            raise error.TestError("%s" % output)
+        if not output:
+            raise error.TestError("Didn't find files in %s, pls check"
+                        "if the driver folder is existing" % driver_path)
+        inf_file = output.split()[0]
+        inf_file = "%s\%s" % (driver_path, inf_file)
+
+        logging.info("Get driver version from %s" % inf_file)
+        output = session.cmd("type %s" % inf_file)
+        ver_pattern = "DriverVer=(.*)"
+        version = re.findall(ver_pattern, output)[0]
+        driver_ver = int(version.split(".")[-1])
+
+        return (driver_path, driver_ver)
+
+    def driver_install(driver_path, session):
+        """
+        Install driver and copy driver debug log to host
+
+        :param driver_path: path of the driver
+        """
+        driver_install_cmd = params.get("driver_install_cmd")
+        winutils = utils_misc.get_winutils_vol(session)
+        install_cmd = driver_install_cmd % (winutils, driver_path)
+        status, output = session.cmd_status_output(install_cmd, timeout=120)
+        get_installation_log(vm)
+
+    def check_driver_version(session):
+        device_name = params.get("device_name")
+        driver_ver_cmd = params.get("driver_version_cmd") % device_name
+        s, output = session.cmd_status_output(driver_ver_cmd)
+        if s:
+            raise error.TestFail(output)
+        driver_version = output.split()[-1].split(".")[-1]
+        return int(driver_version)
+
+    vm = env.get_vm(params["main_vm"])
+    vm.verify_alive()
+    session = vm.wait_for_login()
+    driver_name = params.get("driver_name")
```

Comment:
driver_name used in below, so better to use params["driver_name"], if not set test will failed ASAP.